server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -199,13 +201,18 @@ private boolean hasNewNodes(DiscoveryNodes nodes) {
         return false;
     }
 
-    class InternalAsyncFetch<T extends BaseNodeResponse> extends AsyncShardFetch<T> {
+    class InternalAsyncFetch<Response extends BaseNodeResponse, Cached> extends AsyncShardFetch<Response, Cached> {
 
         InternalAsyncFetch(Logger logger, String type, ShardId shardId, String customDataPath,
-                           Lister<? extends BaseNodesResponse<T>, T> action) {
+                           Lister<? extends BaseNodesResponse<Response>, Response> action) {
             super(logger, type, shardId, customDataPath, action);
         }
 
+        @Override
+        protected Cached extract(Response value) {
+            return null;
```

Comment:
That looks suspicious: why are we not caching anything here now?

Reply:
Since we don't know the exactly target tranfered cache object here. It would be overrided in `InternalPrimaryShardAllocator.fetchData` and `InternalReplicaShardAllocator.fetchData`.

Reply:
I see, could we make `GatewayAllocator` into an abstract class instead?

Reply:
Sorry I didn't get your point. What's the purpose to make `GatewayAllocator` as an abstract class?

Reply:
You are claiming that this `return null` is dead code, never executed, because all subclasses override this method. Making the class `abstract` would let the compiler check this, and would mean that we can delete this dead code.

Reply:
OK, I got it. That should be `InternalAsyncFetch` class which is the inner class, not `GatewayAllocator`.

Reply:
Ah sorry you're right I was just reading this on Github so I didn't notice the inner class there. Thanks.